Agnes distance

targos opened the pull request.

targos commented:
Ref: #4

targos commented on .travis.yml:
```diff
@@ -1,4 +1,4 @@
-  - "iojs"
\ No newline at end of file
+  - "iojs"
+  -"stable"
```
you can remove iojs and instead put:
- 4
- 6
- stable

targos commented on src/agnes.js:
```diff
@@ -153,11 +153,10 @@ function agnes(data, options) {
         d = {};
         min = 10e5;
-        for (var j = 0; j < list.length; j++)
+        for (var j = 0; j < list.length; j++){
```
style: space before the brace

targos commented:
The commit messages should start with a lowercase. It should also use the imperative tense and the first line must be small. The rest of the explanations go on the next lines. Example:

Instead of

```
Bug: Dependency of ml-stat was removed to keep the package small and to keep compatibility with versions of node.
```

Write something like:

```
fix: remove dependency on ml-stat

This keeps the package small and keeps compatibility with versions of node.
```

targos commented:
You can change the commit messages with the `git rebase` command

andcastillo commented:
I commit the messages wrongly again. I'm trying the rebase

andcastillo commented:
@targos I've finished my changes on the agnes code. The git rebase was cumbersome, and I don't know if I get the correct result. I see new commits but my aim was only to change the commit messages.

andcastillo commented on src/agnes.js:
```diff
@@ -83,43 +73,59 @@ function centroidLink(cluster1, cluster2, disFun) {
     disFunc: euclidean,
-    kind: 'single'
+    kind: 'single',
+    source:'data'
```
@targos @maasencioh What do you think about the name of the option that specify that the input is a distance matrix??

maasencioh replied:
I think that we have two options, to call it `'matrix'` or to put the matrix value in there

targos replied:
I would do it like in [ml-pca](https://github.com/mljs/pca/blob/5e1b8f361bdba20d9032cb6dab638f741fe8d142/src/pca.js#L44-L47) and have an option like `isDistanceMatrix: true`

andcastillo commented on src/agnes.js:
```diff
@@ -83,43 +73,59 @@ function centroidLink(cluster1, cluster2, disFun) {
     var len = data.length;
 
+    var distance = data;//If source
+    if(options.source === 'data' ) {
```
@mljs/collaborators How to deal with the options. In this case, if "source" is set to "data", the input is an array of entries, but if "source" is set to "distance" the input is a distance matrix. But in this case, if source is something different than "data", then it is considered as a distance matrix.  Should I consider the else if(option.source === 'distance') case and throw an error if the options source is not valid??

targos committed:
chore: update travis config

targos committed:
feat(agnes): add agglomerative clustering algorithm based on a distance matrix

It can replace the current implementation of agnes, just calculating the distance matrix from the data input at the beginning.
Also, some errors in the code were fixed and documentation updated.

targos commented:
@andcastillo 
I have no idea how you arrived to this result with the rebase (there should not be a merge commit, rebase is kind of the opposite of merge). I restored the branch as it should be and changed the commit messages.
I did a screencast of this, that you can find here: https://asciinema.org/a/d3tglfot8fek4j0mx3k7xme9t

Since I completely rewrote the history of the branch, you need to hard reset it on your side before continuing the work:
- `git checkout agnes-distance`
- `git fetch origin`
- `git reset --hard origin/agnes-distance`

targos commented:
Here is some information about history rewriting: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

andcastillo committed:
isDistanceMatrix replace the source option

andcastillo commented:
@targos I changed the 'source' option by your suggestion 'isDistanceMatrix'. Can we merge and release a new version of the lib?

targos commented:
@andcastillo done, thanks for your work
